### ISSUE TITLE
Computing forward gradients in a compressed manner.

### ISSUE THREAD

**Issue by eneftci** [google/jax#4516]
I'm interested in implementing a approximate, sparse form of forward AD, where some directions are ignored in the jvp. This is motivated by the fact that the gradient of, say, f(x) = W@X wrt W is going to have many zeros on the off-diagonal terms. To reduce memory usage, we can reduce the matrix to a dense one and represent only the non-zero terms. (In a more complex function, these will not be zeros so dropping them introduces a bias in the final gradients, that is intentional and part of the approximation). 

**The question is the following: can I implement a custom jvp function to capture this behavior?** 

Here is how I can hack it in a custom jacfwd. 
`
from jax import vmap
from jax import linear_util as lu
from jax.api import tree_map, argnums_partial, _unravel_array_into_pytree, _check_output_dtype_jacfwd, _std_basis, _check_input_dtype_jacfwd, _jvp

def our_jacfwd(fun, argnums = 0, holomorphic = False):
  def jacfun(*args, **kwargs):
    f = lu.wrap_init(fun, kwargs)
    f_partial, dyn_args = argnums_partial(f, argnums, args)
    tree_map(partial(_check_input_dtype_jacfwd, holomorphic), dyn_args)
    pushfwd = partial(_jvp, f_partial, dyn_args)
    s = _std_basis(dyn_args)[0]
    s = s[::dyn_args[0].shape[1],:,0] #custom tangent space
    print(dyn_args[0].shape)
    y, jac = vmap(pushfwd, out_axes=(None, -1))((s,))
    print(jax.numpy.array(_std_basis(dyn_args)).shape)
    tree_map(partial(_check_output_dtype_jacfwd, holomorphic), y)
    example_args = dyn_args[0] if isinstance(argnums, int) else dyn_args
    print(example_args)
    ##TODO: reshape
    #tree_map(partial(_unravel_array_into_pytree, example_args, -1), jac)
    return y,jac 

  return jacfun`

The idea here is to pick the tangent space of interest. Rewriting jacfwd is not optimal however, because we may not want to use this approximation across the entire graph. I think this is related to the bitrotted elementwise_jvp and jarrett decorator discussed here #4024 

Thanks,
Emre

**Comment by shoyer** [comment]
One other option to reduce memory usage would be to replace the `vmap` inside `jacfwd` with `lax.map`, which uses a loop. In the long term, I suspect we want to use some version of `gmap` from `jax.experimental.general_map` inside `jacfwd`, rather than hard-coding `vmap`.

We definitely want to enable experimentation in JAX around alternative auto-diff implementation like this, but I'm not sure how feasible it would be to allow "swapping out" JAX's internal JVP implementation.

A more realistic option might be trying to reduce the boilerplate inside functions like `jacfwd`, so you can realistically write your own version of JAX's auto-diff system in just a handful of lines of code. Tools like the experimental (unmerged) `tree_vectorize` could in principle help here: https://github.com/google/jax/pull/3263#issuecomment-640252077
